### PR TITLE
Retriever classes

### PR DESCRIPTION
**Pull request by dellaert**
@womackj1 and I split the big PR, so NetVlad is easier to review @akshay-krishnan  :-)

The base retriever class and two instantiations, with tests.

(Copied from netvlad branch, who will now target this branch)

### CODE REVIEW

File: gtsfm/runner/gtsfm_runner_base.py [hunk 1]
```diff
@@ -86,20 +96,39 @@ def construct_scene_optimizer(self) -> SceneOptimizer:
...
+        """Run the SceneOptimizer."""
         start_time = time.time()
+
+        # create dask client
+        cluster = LocalCluster(
+            n_workers=self.parsed_args.num_workers, threads_per_worker=self.parsed_args.threads_per_worker
+        )
+
+        pairs_graph = self.retriever.create_computation_graph(self.loader)
+        with Client(cluster), performance_report(filename="dask-report.html"):
```
Comment (dellaert) [inline gtsfm/runner/gtsfm_runner_base.py:121]: @johnwlambert why do we now have a second `with Client` clause here? The compute only happens in line 135, right?

File: tests/retriever/test_sequential_retriever.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+DATA_ROOT_PATH = Path(__file__).resolve().parent.parent / "data"
+DEFAULT_FOLDER = DATA_ROOT_PATH / "set1_lund_door"
+
+
+class TestSequentialRetriever(unittest.TestCase):
+    def test_sequential_retriever(self) -> None:
+        """Assert that we get 30 total matches with a lookahead of 3 frames on the Door Dataset."""
+
+        max_frame_lookahead = 3
+        loader = OlssonLoader(str(DEFAULT_FOLDER), image_extension="JPG", max_frame_lookahead=max_frame_lookahead)
```
Comment (dellaert) [inline tests/retriever/test_sequential_retriever.py:21]: @johnwlambert why do we still need this lookahead parameter now?

**Comment by dellaert** [comment]
Just changed (sole) reviewer of this to @travisdriver , so @akshay-krishnan can concentrate on the Netvlad one (which is less of a priority, BTW).

File: gtsfm/runner/gtsfm_runner_base.py (hunk 1 above)
Comment (johnwlambert) [reply gtsfm/runner/gtsfm_runner_base.py:121]: Hi @dellaert, the graph is now partitioned into 2 separate computation graphs.

There is a `.compute()` call on line 122, and also on 135. Dask cannot iterate over a sequence of unknown length, which the image pair indices end up being after the Retriever.

I don't think there is any way around this, unfortunately, in our Dask paradigm.

File: tests/retriever/test_sequential_retriever.py (hunk 2 above)
Comment (johnwlambert) [reply tests/retriever/test_sequential_retriever.py:21]: We need it because sequential matching needs a specification of the sliding window size.

**Review by johnwlambert — APPROVED** [review]
LGTM

File: tests/retriever/test_sequential_retriever.py (hunk 2 above)
Comment (dellaert) [reply tests/retriever/test_sequential_retriever.py:21]: Yes, but why is that not just a retriever parameter?

**Comment by dellaert** [comment]
All, where can I find that "overview report" in the CI, showing green/red grid?

File: tests/retriever/test_sequential_retriever.py (hunk 2 above)
Comment (johnwlambert) [reply tests/retriever/test_sequential_retriever.py:21]: You're right, good catch. Can be removed. I recommend to do it in a small follow-up PR.

**Comment by dellaert** [comment]
I ran comparison. Weird that this change, which is supposed to be neutral, has changes.

image

Full results here: [visual_comparison_dashboard.html.pdf](https://github.com/borglab/gtsfm/files/8645104/visual_comparison_dashboard.html.pdf)

**Comment by dellaert** [comment]
New metrics no longer include skipped artifacts (see #475) and look better (but still not neutral):
image

Still, after email thread I will merge.

**Pull request merged**